### DISCUSSION TITLE
To manage data, would you use duckdb or arrow?

### DISCUSSION THREAD

**Discussion by hope-data-science** [duckdb/duckdb#6072] [General]
I read a post in <https://duckdb.org/2021/12/03/duck-arrow.html>, and wondering if I have like ~300G data warehouse, should I save it as arrow files (like feather or parquet) or duckdb file to gain better performance (consider io speed, file size, process speed).

Thanks.

**Comment by Mytherin** [comment]
Currently I would recommend saving data you intend to keep around for longer as Parquet files. The DuckDB file format is indeed faster to operate on - but we have not yet stabilized the file format, which means that future versions of DuckDB will not be able to read files written by the current version of DuckDB. That means when upgrading the DuckDB version data must be converted to the new file format.

In the future when the file format is stabilized the DuckDB format will become a better option for long-term storage.

**Reply by hope-data-science** [reply]
Thank you for the feedback, but by using Parquet I have no idea how to append data to the exist Parquet. So DuckDB file is the best option to go. If possible, when the data format upgrade, is there a method to upgrade the past data to the new format as well? Thanks.

**Reply by Mytherin** [reply]
Yes, you can upgrade the file using the [`IMPORT` and `EXPORT` commands](https://duckdb.org/docs/sql/statements/export). You can export the data to Parquet using the old version of DuckDB, and import the data back into the new version using the import command. This does involve a rewrite of the table - so if you are storing a lot of data then this might be time consuming.

**Reply by hope-data-science** [reply]
I am using R to process data, if I import it into R session, it can be exported directly to duckdb using APPEND method. It would not be a problem, looking forward to the stable version of duckdb format.

**Reply by FrancoisLepoutre** [reply]
Hi hope-data-science,

An outsider view from a pythonista testing the duck, I found that:
1)loading initial stuff (csv-s) into a duckdb console straight from csv was comfortable enough, a temporay database duckdb is created at that stage, 
2) time for some additional SQL viewing mangling on this temp database within the duckdb console, a great tool;
3) exporting from duckdb into a series of parquets for persistency - the COPY to parquet is dammed flexible, I'd recommend SNAPPY,
4) dropping the the temp database.

Now we are in a position where analytical work is simple as it just require to load our parquets from an in-memory database and work from there. 

As Mark said, let us work this way until the duckdb format is stabilized. As the parquet data load really shines, that might even be a permanent solution for some of us :-)

**Reply by hope-data-science** [reply]
This takes extra time to convert csv or parquet to duckdb, which is not desirable. And when we choose a database, we actually have a pretty large dataset, using parquet could be hard for fast query (if you have any good solutions to do fast queries in parquets out-of-disk, let me know). Therefore, looking forward to the later version of duckdb, hope it will be small in size and fast to read and write~ :-)

**Reply by FrancoisLepoutre** [reply]
From the Duckcon that I attended in Brussels on Friday last week - a jolly good time - I understand that the final db format will take care of the issues you mentioned. The format is intended to become a viable alternative to bunches of parquets, possibly large ones. But yes, as Mark said, we have to wait as they intend to deliver once it can effectively compete!

> This takes extra time to convert csv or parquet to duckdb, which is not desirable.

I certainly agree that it takes time to load csv-s into duckdb. But loading "snappy parquets" here was snappy enough to justify working with an in-memory duckdb database as a temp solution. But your mileage may vary depending on needs and database size. Ours ain't that big by the way. 300 G data looks a bit big in my view for our to-day's duck. But I have no experience of such data set sizes anyway.

Currently the duckdb engine is highly effective as long as the data "mostly" fits in memory. And there is definitely a relatively large penalty when the engine is short  of mem. But I was not surprised in view of the technology. But, sure depending on the size of the parquet stuff, loading data from scratch may become difficult! By the way that's an issue the team is specifically trying to tackle as well.

I am not currently looking for alternatives as we are still in a POC phase and this duck fits our specific bill. Just waiting for the project to deliver what they have promised. Up to to-day, the team certainly did deliver. So I'm confident that this situation will be fully cleared in the 1.0 version :-)
